File: pegen/Java/Adaptor.java
import org.antlr.v4.runtime.CharStream;
import org.antlr.v4.runtime.Lexer;
import org.antlr.v4.runtime.Token;

public abstract class Adaptor extends Lexer {

    protected Adaptor(CharStream input) {
	super(input);
    }

    private int CurrentRuleType = Token.INVALID_TYPE;

    public boolean AtEnd()
    {
	return this._mode == pegen_v3_10Lexer.ACTION_MODE;
    }

    protected void handleEndAction()
    {
	int oldMode = this._mode;
	int newMode = this.popMode();
	boolean isActionWithinAction = !this._modeStack.isEmpty()
				    && newMode == pegen_v3_10Lexer.ACTION_MODE
				    && oldMode == newMode;

	if (isActionWithinAction)
	{
	    CurrentRuleType = (pegen_v3_10Lexer.ACTION);
	}
    }
}

File: pegen/CSharp/Adaptor.cs
using Antlr4.Runtime;
using Antlr4.Runtime.Misc;
using System;
using System.IO;
using System.Reflection;

public abstract class Adaptor : Lexer
{
    readonly ICharStream stream;
    readonly FieldInfo tokenInput = typeof(CommonToken).GetField("_type", BindingFlags.NonPublic | BindingFlags.Instance);
    protected Adaptor(ICharStream input)
            : base(input, Console.Out, Console.Error)
    {
        stream = input;
    }

    protected Adaptor(ICharStream input, TextWriter output, TextWriter errorOutput)
            : base(input, output, errorOutput)
    {
        stream = input;
    }

    private int CurrentRuleType { get; set; } = TokenConstants.InvalidType;

    public bool AtEnd()
    {
        return CurrentMode == pegen_v3_10Lexer.ACTION_MODE;
    }

    protected void handleEndAction()
    {
        int oldMode = CurrentMode;
        int newMode = PopMode();
        bool isActionWithinAction = ModeStack.Count > 0
            && newMode == pegen_v3_10Lexer.ACTION_MODE
            && oldMode == newMode;

        if (isActionWithinAction)
        {
            CurrentRuleType = (pegen_v3_10Lexer.ACTION);
        }
    }
}
